### PR TITLE
Fetch non-cached reports in parallel

### PR DESCRIPTION
**Pull request by jml**
Currently, if a bunch of requested reports aren't in the cache, we fetch them one at a time inline with the request. This patch updates the code to fetch reports in parallel instead.

While making the change, I noticed that we handled S3 errors differently from gzip & decoder errors. I've updated the code to handle all errors in the same way (i.e. no special logging, just return).

Currently, `getNonCached` returns only the first error it finds. It should probably aggregate all of the errors somehow so we aren't silently ignoring important conditions. I don't know how best to do that.

Along the way, I extracted the logic for fetching a single report (to make the actual parallelism logic more clear).

---
This change is [Reviewable](https://reviewable.io/reviews/weaveworks/scope/1554)

### CODE REVIEW

File: app/multitenant/dynamo_collector.go [hunk 1]
```diff
@@ -240,37 +240,61 @@ func (c *dynamoDBCollector) getCached(reportKeys []string) ([]report.Report, []s
...
+			r.report, r.err = c.getNonCachedReport(reportKey)
+			ch <- r
+		}(reportKey)
+	}
+
+	reports := []report.Report{}
+	for range reportKeys {
+		r := <-ch
+		if r.err != nil {
+			return nil, r.err
```
Comment (tomwilkie) [inline app/multitenant/dynamo_collector.go]: I think this will leave orphaned goroutines if the first response contains an error.

Comment (tomwilkie) [inline app/multitenant/dynamo_collector.go]: Actually maybe not.

**Comment by tomwilkie** [comment]
Did you see any performance improvement running this on your laptop?

**Comment by jml** [comment]
Haven't actually run it on my laptop, so will at least do that before landing.

I'll also at least attempt to write unit tests (but might end up deciding they're too hard).

File: app/multitenant/dynamo_collector.go [hunk 2]
```diff
@@ -240,37 +240,61 @@ func (c *dynamoDBCollector) getCached(reportKeys []string) ([]report.Report, []s
...
 		}
-		reports = append(reports, rep)
-		c.cache.Set(reportKey, rep)
+		reports = append(reports, *r.report)
+		c.cache.Set(r.key, *r.report)
```
Comment (jml) [inline app/multitenant/dynamo_collector.go]: Thanks, I wasn't sure so erred on the side of caution. Out of curiosity, how could I have inferred that it was safe without asking you?

Comment (tomwilkie) [inline app/multitenant/dynamo_collector.go]: Read the docs: https://godoc.org/github.com/bluele/gcache

**Comment by jml** [comment]
Ran this locally & verified that it doesn't break. No metrics. Won't bother with unit tests now.

@tomwilkie OK to merge

**Comment by jml** [comment]
@tomwilkie asked for metrics:

### Before

20160608-143026_3196x1717_scrot

### After

20160608-144305_3196x1717_scrot

Running on my laptop locally.

**Comment by tomwilkie** [comment]
LGTM

**Pull request merged**